Parse "dir" member of the web application manifest

christianliebel opened the pull request:
#### 1b2b03786fb750eb56bed4e2d64d8477f8552f1e
<pre>
Parse &quot;dir&quot; member of the web application manifest
<a href="https://bugs.webkit.org/show_bug.cgi?id=276409">https://bugs.webkit.org/show_bug.cgi?id=276409</a>
<a href="https://rdar.apple.com/problem/131900106">rdar://problem/131900106</a>

Reviewed by Darin Adler.

Implements the &quot;dir&quot; member to the web application manifest parser per spec:
<a href="https://www.w3.org/TR/appmanifest/#dir-member">https://www.w3.org/TR/appmanifest/#dir-member</a>
Along with the required encorders/decoders and API tests.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseDir):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest direction]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(WebCore::operator&lt;&lt;):
(ApplicationManifestParserTest::testDir):
(TEST_F):

Canonical link: <a href="https://commits.webkit.org/282761@main">https://commits.webkit.org/282761@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e39a68944aa87c8c36df7d28ecc1c5f63071f6df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64052 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43409 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16649 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68074 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14660 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66172 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51107 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14940 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51584 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10121 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40174 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55432 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32203 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36846 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12817 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13533 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58804 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13146 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69773 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7999 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12671 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58903 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8032 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55530 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59054 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14174 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6631 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39229 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40308 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41491 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40051 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres reviewed (approved):
Non-reviewer approved. Thanks @christianliebel!

annevk commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -173,6 +174,32 @@ URL ApplicationManifestParser::parseStartURL(const JSON::Object& manifest, const
+    };
+    static constexpr SortedArrayMap directions { directionMappings };
+
+    if (auto* dirValue = directions.tryGet(StringView(stringValue).trim(isUnicodeCompatibleASCIIWhitespace<UChar>)))
```
Using `isUnicodeCompatibleASCIIWhitespace` is almost certainly a bug. What kind of whitespace does the specification call for?

christianliebel replied:
I have adopted this line of code to ensure consistency with the parsing of [display modes](https://github.com/WebKit/WebKit/blob/044553fa65f400aee0001dcf2177cf78b49a53f5/Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp#L223) and [orientations](https://github.com/WebKit/WebKit/blob/044553fa65f400aee0001dcf2177cf78b49a53f5/Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp#L255). The trimming behavior is also covered by tests ([example](https://github.com/WebKit/WebKit/blob/044553fa65f400aee0001dcf2177cf78b49a53f5/Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp#L496)). The specification does not provide guidance on handling whitespace.

marcoscaceres replied:
Yeah, the spec doesn't trim anymore... we did some archeology... the spec used to trim back in 2017, but then it got dropped when we converted things to WebIDL. 

Discussed with @annevk.... we probably want to do this here:
 
```suggestion
    if (auto* dirValue = directions.tryGet(StringView(stringValue).trim(isASCIIWhitespace<UChar>)))
```

marcoscaceres replied:
Filed bug for the others... https://bugs.webkit.org/show_bug.cgi?id=278377

annevk commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
```diff
@@ -36,6 +36,12 @@
+    enum class Direction : uint8_t {
+        Auto,
+        Ltr,
+        Rtl,
```
Shouldn't the casing be LTR and RTL?

christianliebel replied:
That would be great, but then the linter complains and wants to have either `kLTR` or `Ltr`.

annevk replied:
It's okay to ignore the linter when it is wrong, although https://webkit.org/code-style-guidelines/#names-basic does not making a ruling either way. As I'd like someone else to review anyway, hopefully they can look at this as well.

marcoscaceres replied:
I flagged the same thing... we might just need to disable the linter for this. That's ok.

darinadler replied:
To be more precise, we often code exceptions into the linter for cases like these, rather than "disabling" it; I think it’s pretty easy to add one and that change can be included in the pull request. By the way, for such a short list I often suggest using a single-line format, but I guess that’s not happening below for `Display`.

marcoscaceres replied:
Yes, sorry for not being clear. By "disabling it" I meant a one line exception.

marcoscaceres replied:
For reference if anyone lands here:  `git grep "// NOLINT"`

darinadler commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -173,6 +174,32 @@ URL ApplicationManifestParser::parseStartURL(const JSON::Object& manifest, const
+{
+    auto value = manifest.getValue("dir"_s);
+    if (!value)
+        return ApplicationManifest::Direction::Auto;
```
Can put `using enum ApplicationManifest::Direction` at the top of the function and make the entire thing more terse without losing clarity.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
```diff
@@ -364,6 +367,20 @@ - (NSString *)rawJSON
 
+- (_WKApplicationManifestDirection)direction
+{
+    switch (_applicationManifest->applicationManifest().dir) {
```
Can put `using enum ApplicationManifest::Direction` at the top of the function and make the entire thing more terse without losing clarity.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -36,6 +36,18 @@
 namespace WebCore {
+static inline std::ostream& operator<<(std::ostream& os, const ApplicationManifest::Direction& direction)
+{
+    switch (direction) {
```
Can put `using enum ApplicationManifest::Direction` at the top of the function and make the entire thing more terse without losing clarity.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -444,6 +463,23 @@ TEST_F(ApplicationManifestParserTest, StartURL)
 
+TEST_F(ApplicationManifestParserTest, Dir)
+{
+    testDir("123"_s, ApplicationManifest::Direction::Auto);
```
Can put `using enum ApplicationManifest::Direction` at the top of the function and make the entire thing more terse without losing clarity.

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
```diff
@@ -170,6 +182,13 @@ class ApplicationManifestParserTest : public testing::Test {
+    {
+        auto manifest = parseTopLevelProperty("dir"_s, rawJSON);
+        auto value = manifest.dir;
+        EXPECT_EQ(expectedValue, value);
```
Why not this one-liner?
```
EXPECT_EQ(expectedValue, parseTopLevelProperty("dir"_s, rawJSON).dir);
```
Doesn’t seem like the local variables help us here.

darinadler commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
```diff
@@ -36,6 +36,12 @@
 struct ApplicationManifest {
+    enum class Direction : uint8_t {
+        Auto,
+        LTR, // NOLINT
+        RTL, // NOLINT
```
This is fine for now. I think exceptions hard-coded into the linter script for "Ltr" and "Rtl" would be better long-term.

christianliebel replied:
I filed PR #32598 to add "LTR" and "RTL" to the allowlist.

marcoscaceres replied:
Cool! We could even fix the NotificationDirection enum. 🎉 (as a follow up!)

darinadler commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -173,6 +174,34 @@ URL ApplicationManifestParser::parseStartURL(const JSON::Object& manifest, const
+    if (auto* dirValue = directions.tryGet(StringView(stringValue).trim(isASCIIWhitespace<UChar>)))
+        return *dirValue;
+
+    logDeveloperWarning(makeString("\""_s, stringValue, "\" is not a valid dir."_s));
```
It would be more efficient to use `'"'` instead of `"\""_s`, which would otherwise accomplish the same thing; some day maybe we can figure out how to get the compiler to warn about this or perhaps compile it into the same code to eliminate the efficiency difference.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
```diff
@@ -267,6 +267,7 @@ + (BOOL)supportsSecureCoding
 - (instancetype)initWithCoder:(NSCoder *)aDecoder
 {
     String rawJSON = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"raw_json"];
+    NSInteger dir = [aDecoder decodeIntegerForKey:@"dir"];
```
I think for security reasons we would want this to check for integers that are out of range, and fail to decode in that case instead of making an illegal enumeration value.

marcoscaceres replied:
The manifest parser should guard for this , so would it be ok to just?:

```suggestion
    NSInteger dir = [aDecoder decodeIntegerForKey:@"dir"];
    // FIXME: <https://webkit.org/b/278619> Remove this assert after further manifest IPC hardening.
    RELEASE_ASSERT(dir >= 0 && dir <= 2);
```

darinadler replied:
What guarantees this was encoded by the manifest parser?

marcoscaceres replied:
Statically, that one can only get here by having gone through the manifest parser machinery and associated call sites (coming in from `HTMLLinkElement` being responsible for loading the manifest and having processed it, `Document` having provided it... and so on). 

I get your point that there's no guarantees computationally that how we got here was from the route above (something else could have IPC'ed its way here), but that then starts applying equally to just about anything that sends data over IPC to anything in the UI Process. There's a presupposition of trust between the UI Process and WebCore because the generated code from the .in file validated and guaranteed the `dir`'s range:

```
[Nested] enum class WebCore::ApplicationManifest::Direction : uint8_t {
    Auto
    LTR
    RTL
};
```

The presumption for the entire _WKApplicationManifest.mm file is that the data was processed through the manifest parser, and that the manifest parser conform the W3C spec, which is a strict validating parser/processor, and then the IPC layer only allowed `WebCore::ApplicationManifest::Direction` to pass through. If that invariant was to fall apart (i.e., we don't trust our own IPC layer), we would need recheck everything.

marcoscaceres replied:
Spoke with @geoffreygaren. Will check if we need to sprinkle MESSAGE_CHECK a bit closer to the IPC layer for the entire parsed manifest structure. 

I've filed a bug to do that as followup so we can do that all at once: 
https://bugs.webkit.org/show_bug.cgi?id=278619

christianliebel replied:
@marcoscaceres I have included your suggestion in my last push.

christianliebel committed:
Parse "dir" member of the web application manifest
https://bugs.webkit.org/show_bug.cgi?id=276409
rdar://problem/131900106

Reviewed by Darin Adler.

Implements the "dir" member to the web application manifest parser per spec:
https://www.w3.org/TR/appmanifest/#dir-member
Along with the required encorders/decoders and API tests.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseDir):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest direction]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(WebCore::operator<<):
(ApplicationManifestParserTest::testDir):
(TEST_F):

Canonical link: https://commits.webkit.org/282761@main
